### PR TITLE
RTDETRv1/v2 Bounty

### LINKED ISSUE #30869: [Bounty $1500] RT-DETR bring up using TTNN APIs
### 📝 Background

This bounty is for bringing up [RT-DETR (Real-Time DEtection TRansformer)](https://github.com/lyuwenyu/RT-DETR) using TTNN APIs on Tenstorrent hardware (Wormhole or Blackhole). RT-DETR is an end-to-end transformer-based object detector designed for real-time inference. It eliminates non-maximum suppression (NMS) and uses a hybrid CNN–Transformer encoder with dynamic query selection to balance speed and accuracy. Later versions (e.g., RT-DETRv2) introduce bag-of-freebies training improvements and architecture refinements for even higher performance.  

The goal is to enable efficient, end-to-end RT-DETR inference on TT hardware for high-throughput, low-latency object detection, showcasing the platform’s suitability for transformer-heavy vision workloads.

### 🎯 What Success Looks Like

A successful submission will fulfill all requirements below. Payout is made after **all three stages** are completed.

### Stage 1 — Bring-Up

- Implement RT-DETR using TTNN APIs (Python).  
- Runs on N150 or N300 without runtime errors.  
- Produces valid detections (bounding boxes + class labels) on sample images.  
- Output is numerically/verifiably aligned with a PyTorch reference (small validation set).  
- Accuracy evaluated on the COCO dataset, following [YOLOv4 demo instructions](https://github.com/tenstorrent/tt-metal/tree/main/models/demos/yolov4#coco-2017-dataset).  
- Include clear environment setup and run instructions.

### Stage 2 — Basic Optimizations

- Use optimal sharded/interleaved memory configs for convolutional and attention layers.  
- Fuse simple ops where possible (e.g., Linear + Bias + Activation, or conv + activation).  
- Store intermediates in L1 when beneficial.  
- Follow recommended TTNN/tt-metal CNN + Transformer flows.  
- Use TT fused ops like [relu with conv](https://github.com/tenstorrent/tt-metal/pull/29236).  
- Profile encoder/decoder stages and eliminate redundant Tensor Manipulations (TMs).

### Stage 3 — Deeper Optimization

- Maximize core parallelism across encoder/decoder layers.  
- Implement deeper TT-specific optimizations (e.g., fused QKV + SDPA, operator reordering to reduce TM overheads).  
- Optimize hybrid CNN–Transformer fusion to minimize memory transfers.  
- Document any advanced tuning or known issues.

### 🧭 Guidance & Starting Points

- Use the [TTNN model bring-up tech report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md) as your primary reference.
- Reference [Yolov4 implementation (PR 29157)](https://github.com/tenstorrent/tt-metal/pull/29157) for example end-to-end flows.
- [RT-DETR Paddle/PyTorch implementation](https://github.com/lyuwenyu/RT-DETR)
- Refer to [TT fused ops](https://github.com/tenstorrent/tt-metal/pull/29236) for opportunities to optimize.
- Target input resolutions: Start with 640×640 images (common config) and COCO-style evaluation; adjust as needed for TT hardware.
- Ask for help or file issues if ops are missing in TTNN.

### 🔎 Possible Approaches

- Port an existing PyTorch RT-DETR step-by-step (backbone → encoder → decoder → prediction heads).  
- Validate intermediate tensors against CPU/PyTorch reference.  
- Profile encoder attention and multi-scale fusion; apply fusion/reshape optimizations to reduce TM overheads.  
- Add decoder-layer control flags to demonstrate speed vs. accuracy scaling.  
- Use TTNN profiling tools to identify bottlenecks.  
- Open a draft PR early for feedback on approach and performance metrics.

### 📊 Result submission guidelines notes

Beyond the model implementation itself. Contributors must submit the following material as a proof of work. However, feel free to open a PR at any time if you want us checking you are on the right track. Just understand that payout is only made after all 3 stages are completed.

- Functional model implementation
- Validation logs (output correctness)
- Performance report + header for final review

Links:
- [Link to perf sheet](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md#41-performance-sheet)
- [Perf header](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/profiling_ttnn_operations.html#perf-report-headers)

---

### 📚 Resources

- [TTNN model bring-up tech report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md)
- [CNN Bring-up & Optimization in TT-NN](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/CNNs/cnn_optimizations.md)
- [Yolov4 TTNN implementation PR](https://github.com/tenstorrent/tt-metal/pull/29157)
- [RT-DETR Paddle/PyTorch implementation](https://github.com/lyuwenyu/RT-DETR)
- [TT fused ops PR](https://github.com/tenstorrent/tt-metal/pull/29236)
- [Perf report header](https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/profiling_ttnn_operations.html#perf-report-headers)

### PR DESCRIPTION
**Pull request by ajumpa**
# RTDetr v1/v2 bring up
Bring up of bounty #30869 

### Summary
RTDetrv1/v2 are vision detection models. This PR includes model files, unit tests and a demo.

Focusing on the rtdetr_50vd and rtdetrv2_50vd model variants.

### Notes for reviewers
Changes in`models/demos/vision/detection/rtdetr`  

### Stage 1 — Bring-Up

- Implemented RT-DETR and RT-DETRv2 inference using TTNN Python APIs.
- Validated execution on P150a without runtime errors.
- Added image and video demos with bounding boxes, labels, and confidence scores.
- Added module-level and end-to-end PCC tests against PyTorch.
- Added detection validation using COCO validation images.
- Added preprocessing and postprocessing with `RTDetrImageProcessor`.
- Added demo, correctness-test, and performance-test instructions.

### Stage 2 — Basic Optimizations

- Folded frozen batch normalization into convolution weights and biases.
- Fused ReLU and SiLU into convolution operations.
- Fused activations into residual additions and linear operations.
- Used height-sharded and block-sharded convolution layouts.
- Enabled convolution resharing when necessary.
- Used mixed DRAM/L1 residency for deformable attention.
- Cached prepared convolution weights and biases in device DRAM.
- Deallocated temporary feature tensors after use.
- Retained only required backbone feature maps.
- Precomputed anchors, valid masks, spatial shapes, and level-start indices.
- Cached sinusoidal positional embeddings.
- Enabled TTNN program caching and device trace replay.
- Profiled backbone, encoder, proposal-generation, and decoder stages.
- Replaced the original top-k path with `topk_large_indices` and slicing.
- Replaced expanded gathers with `ttnn.embedding`.
- Implemented deformable attention with pure-TTNN per-level `grid_sample`.
- Removed unused inference branches and redundant runtime work.

### Stage 3 — Deeper Optimizations

- Distributed convolution computation using sharded configurations.
- Used TTNN scaled dot-product attention for transformer self-attention.
- Applied mixed DRAM/L1 residency to reduce deformable-attention transfers.
- Reordered and simplified proposal selection to remove the top-k bottleneck.
- Specialized execution for fixed-shape, batch-1 inference.
- Reduced full-model trace latency from approximately `38.7 ms` to `23.8 ms`.

### Measured Improvements

| Component | Before | After |
|---|---:|---:|
| Proposal generation | ~13.5 ms | ~0.95 ms |
| Decoder | ~19.0 ms | ~17.3 ms |
| Full-model trace | ~38.7 ms | ~23.8 ms |

# Demos 

Outputs saved to `demo/outputs/DATE_ID/`

```
python models/demos/vision/detection/rtdetr/demo/demo.py
```

Or with a specific url or cached image

```
python models/demos/vision/detection/rtdetr/demo/demo.py \
  --image /path/to/image.jpg \
  --threshold 0.5
```

Or

```
python models/demos/vision/detection/rtdetr/demo/demo.py \
  --sample_image [0-9]
```

To use RTDetrv2 model in the demo

```
python models/demos/vision/detection/rtdetr/demo/demo.py \
  --sample_image [0-9] \
  --model_version 2
```

# Validation

## Setup

| Item | Value |
|---|---|
| Hardware | Tenstorrent P150a |
| Git base commit | `b1d4b7a2106b4ebe16f6a42ebcc63755146b3df4` |
| Input | Batch 1, 640×640, BF16 |
| Device configuration | `l1_small_size=16384`, `trace_region_size=1 << 26` |
| RT-DETR v1 checkpoint | `PekingU/rtdetr_r50vd` |
| RT-DETRv2 checkpoint | `PekingU/rtdetr_v2_r50vd` |
| Validation data | First 500 COCO `val2017` images in sorted filename order |

The recorded runs used the listed base commit with uncommitted RT-DETR validation changes.

## COCO Validation Data

The detection tests require the COCO `val2017` images. Ground-truth annotations are not required because the tests compare TTNN detections against the PyTorch reference.

From the RT-DETR directory:

```bash
mkdir -p datasets/coco
cd datasets/coco
wget http://images.cocodataset.org/zips/val2017.zip
unzip val2017.zip
rm val2017.zip
cd ../..
```

The default image path is `./datasets/coco/val2017`. To use another location:

```bash
export COCO_VAL2017_IMAGES=<path-to-val2017>
```

## PCC

```bash
pytest tests/pcc.py::test_rtdetr_model -sv
pytest tests/pcc.py::test_rtdetr_v2_model -sv
```

Decoder queries shared by both implementations were aligned by proposal index before PCC comparison.

| Output | RT-DETR v1 | RT-DETRv2 |
|---|---:|---:|
| Last hidden state | 0.9693 | 0.9699 |
| Intermediate hidden states | 0.9667 | 0.9705 |
| Intermediate logits | 0.9375 | 0.9545 |
| Intermediate reference points | 0.9802 | 0.9512 |

## Detection Correctness

```bash
pytest tests/validate_detection.py::test_rtdetr_detection -sv
pytest tests/validate_detection.py::test_rtdetr_v2_detection -sv
```
Validation on a subset of 500 of the full COCO validation set.

PyTorch and TTNN detections used a confidence threshold of `0.5`. Boxes were greedily matched at IoU ≥ `0.5`.

| Metric | RT-DETR v1 | RT-DETRv2 |
|---|---:|---:|
| Images | 500 | 500 |
| PyTorch detections | 2,763 | 3,550 |
| TTNN detections | 2,752 | 3,558 |
| Matched detections | 2,644 | 3,377 |
| Matched ratio | 95.69% | 94.91% |
| Label agreement | 99.74% | 98.46% |
| Mean matched IoU | 0.9577 | 0.9521 |
| Minimum matched IoU | 0.5196 | 0.5383 |
| Mean confidence-score error | 0.0210 | 0.0221 |

## Model Latency

```bash
pytest tests/latency.py::test_rtdetr_batch_1_trace_latency -sv
pytest tests/latency.py::test_rtdetr_v2_batch_1_trace_latency -sv
```

Each test performs one warmup, captures the model, and measures eight trace replays on different images. Timing includes blocking trace execution and excludes preprocessing, input transfer, output transfer, and postprocessing. Program caching is enabled.

| Metric | RT-DETR v1 | RT-DETRv2 |
|---|---:|---:|
| Mean latency | 23.78 ms | 23.77 ms |
| Median latency | 23.78 ms | 23.78 ms |
| Throughput | 42.05 FPS | 42.07 FPS |

## Video Throughput

```bash
python demo.py --video <path-or-direct-url> --model_version 1
```

Video throughput includes decoding, preprocessing, input transfer, trace replay, output transfer, postprocessing, drawing, and H.264 encoding. Model initialization, warmup, and trace capture are excluded.

| Metric | Result |
|---|---:|
| Mean throughput | 27.83 FPS |
| Minimum throughput | 25.34 FPS |
| Maximum throughput | 29.19 FPS |

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1212806203871061

### CODE REVIEW

**Comment by ajumpa** [comment]
@jberkowitzTT I have a working demo in demos/demo.py. Everything seems to be running pretty well. Some remaining TODOs

- [x] Inference speed optimizations
- [x] RTDetrv2 decoder (Should be a straight forward plug in, just need to familiarize myself with the architecture)
- [x] A detailed README.md with instructions

**Comment by ajumpa** [comment]
Some sample outputs

tt_output0
tt_output1

**Comment by ajumpa** [comment]
v2 model is merged into the demo. There were some trivial alterations to the decoder deformable attention module, the rest of the model architecture is shared. You can run some pcc tests

```
pytest models/demos/vision/detection/rtdetr/tests/tests.py::test_rtdetr_v2_multiscale_deformable_attention -s -v
```

```
pytest models/demos/vision/detection/rtdetr/tests/tests.py::test_rtdetr_v2_model -s -v
```

or the demo

```
python models/demos/vision/detection/rtdetr/demo/demo.py \
    --model_version 2 \
    --sample_image [0-9]
```

**Comment by ajumpa** [comment]
## Sample video demo from the KITTI dataset.

Hits ~25 fps

https://github.com/user-attachments/assets/0e953fd4-11aa-4eae-a3d3-d61428f5b45f

**Comment by ajumpa** [comment]
@jberkowitzTT Im sitting at ~42 fps with the model and ~25fps using it in a demo with postprocessing. If you can address a few questions so I can get this polished up.

1. RT-DETR is for real-time detection, in which case batch size is usually size 1. Do I need to support batch > 1 for this demo?
2. I am using the RT-DETR-r50. There are other variations (larger or smaller). Do I need to support these as well?
3. Apart from what I have above, what other validations or tests do you need to see?

**Comment by jberkowitzTT** [comment]
@ajumpa - Thanks for the update, this looks good. For the bounty scope, batch=1 is sufficient for the demo, and we don’t need the other RT-DETR model sizes. Supporting RT-DETR-R50-vd and RT-DETRv2-R50-vd is enough.

For final validation, please include reproducible N150/N300 run logs, end-to-end correctness results against the PyTorch reference on the COCO validation set for both v1 and v2, and the exact commands and setup used to produce the latency/throughput numbers. The existing PCC, detection, latency tests, and demo instructions cover the rest. Once those logs and the performance methodology are added, I think the validation package is complete.

**Comment by ajumpa** [comment]
@jberkowitzTT Awesome. I do not have access to an N150 or N300, but I do have the p150a which I am using for development. I will get you those items using the p150a, which I think should carry over to the wormhole architecture.

**Comment by ajumpa** [comment]
@jberkowitzTT I added further validation, the commands and setup to reproduce them. Please see the very top comment and let me know what else you will need.

This PR is ready for final review

**Comment by ajumpa** [comment]
@jberkowitzTT CI has passed successfully

**Comment by ajumpa** [comment]
@jberkowitzTT if there are any outstanding items/tasks you or the other code reviewers need for this PR to meet the bounty requirements, please let me know and I will do my best to have them done today.

**Review by uaydonat — APPROVED** [review]
